File: frontend/admin/src/api/conversations.ts
import { fetchWithAuth } from './client'

export interface Conversation {
  id: string
  visitor_id: string
  visitor_external_userid: string
  status: string
  assigned_staff_id: string | null
  transfer_reason: string | null
  created_at: string
  updated_at: string
}

export interface Message {
  id: string
  role: string
  content: string
  msg_type: string
  created_at: string
}

export interface Staff {
  id: string
  username: string
  display_name: string
  wecom_userid: string | null
  is_active: boolean
  created_at: string
  updated_at: string
}

export async function getConversations(status?: string): Promise<Conversation[]> {
  const url = status ? `/api/v1/conversations?status=${status}` : '/api/v1/conversations'
  const r = await fetchWithAuth(url)
  if (!r.ok) throw new Error(`HTTP ${r.status}`)
  return r.json()
}

export async function getConversationMessages(conversationId: string): Promise<Message[]> {
  const r = await fetchWithAuth(`/api/v1/conversations/${conversationId}/messages`)
  if (!r.ok) throw new Error(`HTTP ${r.status}`)
  return r.json()
}

export async function transferConversation(conversationId: string, reason: string): Promise<Conversation> {
  const r = await fetchWithAuth(`/api/v1/conversations/${conversationId}/transfer`, {
    method: 'POST',
    body: JSON.stringify({ reason }),
  })
  if (!r.ok) throw new Error(`HTTP ${r.status}`)
  return r.json()
}

export async function assignStaff(conversationId: string, staffId: string): Promise<Conversation> {
  const r = await fetchWithAuth(`/api/v1/conversations/${conversationId}/assign`, {
    method: 'POST',
    body: JSON.stringify({ staff_id: staffId }),
  })
  if (!r.ok) throw new Error(`HTTP ${r.status}`)
  return r.json()
}

export async function getStaffList(includeInactive = false): Promise<Staff[]> {
  const r = await fetchWithAuth(`/api/v1/staff?include_inactive=${includeInactive}`)
  if (!r.ok) throw new Error(`HTTP ${r.status}`)
  return r.json()
}

export async function createStaff(data: {
  username: string
  display_name: string
  password: string
  wecom_userid?: string
}): Promise<Staff> {
  const r = await fetchWithAuth('/api/v1/staff', {
    method: 'POST',
    body: JSON.stringify(data),
  })
  if (!r.ok) {
    const err = await r.json().catch(() => ({}))
    throw new Error((err as { detail?: string }).detail ?? `HTTP ${r.status}`)
  }
  return r.json()
}

export async function closeConversation(conversationId: string): Promise<Conversation> {
  const r = await fetchWithAuth(`/api/v1/conversations/${conversationId}/close`, {
    method: 'PATCH',
  })
  if (!r.ok) throw new Error(`HTTP ${r.status}`)
  return r.json()
}

export async function updateStaffStatus(staffId: string, isActive: boolean): Promise<Staff> {
  const r = await fetchWithAuth(`/api/v1/staff/${staffId}`, {
    method: 'PATCH',
    body: JSON.stringify({ is_active: isActive }),
  })
  if (!r.ok) throw new Error(`HTTP ${r.status}`)
  return r.json()
}

export async function replyMessage(conversationId: string, content: string): Promise<Message> {
  const r = await fetchWithAuth(`/api/v1/conversations/${conversationId}/reply`, {
    method: 'POST',
    body: JSON.stringify({ content }),
  })
  if (!r.ok) throw new Error(`HTTP ${r.status}`)
  return r.json()
}

File: backend/app/api/v1/conversations.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_current_user
from app.database import get_db
from app.models.staff import Staff
from app.schemas.conversation import (
    AssignStaffRequest,
    ConversationResponse,
    ReplyRequest,
    TransferRequest,
)
from app.services import conversation_service

router = APIRouter(prefix="/conversations", tags=["conversations"])


@router.get("")
async def list_conversations(
    status: str | None = Query(
        default=None, description="按状态过滤: active | transferred | closed"
    ),
    limit: int = Query(default=50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
    _: Staff = Depends(get_current_user),
) -> list[dict[str, object]]:
    return await conversation_service.list_all(db, limit=limit, status=status)


@router.get("/{conversation_id}/messages")
async def get_conversation_messages(
    conversation_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: Staff = Depends(get_current_user),
) -> list[dict[str, object]]:
    msgs = await conversation_service.get_messages(db, conversation_id)
    return [
        {
            "id": str(m.id),
            "role": m.role,
            "content": m.content,
            "msg_type": m.msg_type,
            "created_at": m.created_at.isoformat(),
        }
        for m in msgs
    ]


@router.post("/{conversation_id}/transfer", response_model=ConversationResponse)
async def transfer_conversation(
    conversation_id: uuid.UUID,
    body: TransferRequest,
    db: AsyncSession = Depends(get_db),
    _: Staff = Depends(get_current_user),
) -> ConversationResponse:
    conv = await conversation_service.transfer_conversation(
        db, conversation_id=conversation_id, reason=body.reason
    )
    return ConversationResponse.model_validate(conv)


@router.post("/{conversation_id}/assign", response_model=ConversationResponse)
async def assign_staff_to_conversation(
    conversation_id: uuid.UUID,
    body: AssignStaffRequest,
    db: AsyncSession = Depends(get_db),
    _: Staff = Depends(get_current_user),
) -> ConversationResponse:
    conv = await conversation_service.assign_staff(
        db, conversation_id=conversation_id, staff_id=body.staff_id
    )
    return ConversationResponse.model_validate(conv)


@router.patch("/{conversation_id}/close", response_model=ConversationResponse)
async def close_conversation(
    conversation_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: Staff = Depends(get_current_user),
) -> ConversationResponse:
    conv = await conversation_service.close_conversation(db, conversation_id)
    return ConversationResponse.model_validate(conv)


@router.post("/{conversation_id}/reply")
async def reply_to_conversation(
    conversation_id: uuid.UUID,
    body: ReplyRequest,
    db: AsyncSession = Depends(get_db),
    _: Staff = Depends(get_current_user),
) -> dict[str, object]:
    msg = await conversation_service.reply_and_send_wecom(
        db, conversation_id=conversation_id, content=body.content
    )
    return {
        "id": str(msg.id),
        "role": msg.role,
        "content": msg.content,
        "msg_type": msg.msg_type,
        "created_at": msg.created_at.isoformat(),
    }
